sdk/src/androidTest/java/com/mongodb/stitch/StitchClientTest.kt

```diff
@@ -0,0 +1,218 @@
+    private var stitchClient: StitchClient? = null
+
+    @Before
+    fun setup() {
+        // start the mock server
+        RESTMockServerStarter.startSync(AndroidAssetsFileParser(instrumentationCtx), AndroidLogger())
+
+        // instantiate a new StitchClient using a dummy name and the mock baseUrl
+        val stitchClient = StitchClient(instrumentationCtx, "dummy-app", RESTMockServer.getUrl())
+        this.stitchClient = stitchClient
+
+        // clear all instances of the internal [SharedPreferences] to start with a clean slate
+        stitchClient.properties.clear()
+        stitchClient.javaClass.kotlin.declaredMemberProperties.first {
+            it.name == "_preferences"
+        }.also { it.isAccessible = true }.get(stitchClient).also {
+            (it as SharedPreferences).edit().clear().commit()
+        }
+
+        // mock out all calls related to auth
+        RESTMockServer.whenPOST(pathContains("auth")).thenReturnString(mockAuthData)
+        RESTMockServer.whenGET(pathEndsWith("auth")).thenReturnString(mockProviderData)
+        RESTMockServer.whenDELETE(pathEndsWith("auth")).thenReturnEmpty(200)
+
+        // mock out all calls related to pipelines
+        RESTMockServer.whenPOST(pathEndsWith("pipeline")).thenReturnString(mockPipelineData)
+
+        // mock out user profile call
+        RESTMockServer.whenGET(pathEndsWith("me")).thenReturnString(mockUserData)
+    }
+
+    /**
+     * Test authentication flow
+     */
+    @Test
+    fun testAuth() {
+        var loggedIn = false
+
+        // add an auth listener to the stitchClient, asserting logged in
+        // and logged out status later in this test
+        stitchClient!!.addAuthListener(object : AuthListener {
+            override fun onLogin() {
+                loggedIn = true
+            }
+
+            override fun onLogout(lastProvider: String?) {
+                loggedIn = false
+            }
+        })
+
+        // assert that we have not authenticated yet
+        assertThat(!stitchClient!!.isAuthenticated)
+
+        // fetch mocked authProviders
+        await(stitchClient!!.authProviders).let {
+            // assert all providers provided in mock are available
+            assertThat(it.hasAnonymous() && it.hasEmailPassword()
+                    && it.hasFacebook() && it.hasGoogle())
+
+            // return as an array for convenience
+            arrayOf(it.anonymous, it.emailPassword, it.facebook, it.google)
+        }.forEach {
+            // assert google and facebook auth providers have their
+            // respective ids
+            when (it) {
+                is GoogleAuthProviderInfo ->
+                    assertThat(it.clientId == FakeGoogleClientId)
+                is FacebookAuthProviderInfo ->
+                    assertThat(it.applicationId == FakeFacebookClientId)
+            }
+        }
+
+        // log in anonymously
+        await(stitchClient!!.logInWithProvider(AnonymousAuthProvider()))
```

Comment:
Correct, `!!` is the unsafe way to unwrap optionals. In this case, it's safe as we know 100% that the `stitchClient` will be non-null and unchanged.